From 51643cb290f914251008f8f75707266a4d6f50fb Mon Sep 17 00:00:00 2001
From: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Wed, 30 Nov 2022 10:08:31 -0300
Subject: [PATCH 30/55] core: Introduce base_type__language_defined()

So that we don't emit typedefs for things like 'char', "unsigned int",
etc.

Signed-off-by: Arnaldo Carvalho de Melo <acme@redhat.com>
---
 dwarves.c | 23 +++++++++++++++++++++++
 dwarves.h |  2 ++
 2 files changed, 25 insertions(+)

diff --git a/dwarves.c b/dwarves.c
index b8681bc..b43031c 100644
--- a/dwarves.c
+++ b/dwarves.c
@@ -250,6 +250,29 @@ static struct ase_type_name_to_size {
 	{ .name = NULL },
 };
 
+bool base_type__language_defined(struct base_type *bt)
+{
+	int i = 0;
+	char bf[64];
+	const char *name;
+
+	if (bt->name_has_encoding)
+		name = bt->name;
+	else
+		name = base_type__name(bt, bf, sizeof(bf));
+
+	while (base_type_name_to_size_table[i].name != NULL) {
+		if (bt->name_has_encoding) {
+			if (strcmp(base_type_name_to_size_table[i].name, bt->name) == 0)
+				return true;
+		} else if (strcmp(base_type_name_to_size_table[i].name, name) == 0)
+			return true;
+		++i;
+	}
+
+	return false;
+}
+
 size_t base_type__name_to_size(struct base_type *bt, struct cu *cu)
 {
 	int i = 0;
diff --git a/dwarves.h b/dwarves.h
index 589588e..0518615 100644
--- a/dwarves.h
+++ b/dwarves.h
@@ -1362,6 +1362,8 @@ const char *base_type__name(const struct base_type *btype, char *bf, size_t len)
 
 size_t base_type__name_to_size(struct base_type *btype, struct cu *cu);
 
+bool base_type__language_defined(struct base_type *bt);
+
 struct array_type {
 	struct tag	tag;
 	uint32_t	*nr_entries;
-- 
2.40.0.rc2

